src/decaf.cpp

```diff
@@ -278,6 +278,39 @@ ExecutionResult Interpreter::visitUDiv(llvm::BinaryOperator &op) {
   return ExecutionResult::Continue;
 }
 
+ExecutionResult Interpreter::visitSRem(llvm::BinaryOperator &op) {
+  StackFrame &frame = ctx->stack_top();
+
+  auto lhs = normalize_to_int(frame.lookup(op.getOperand(0), *z3));
+  auto rhs = normalize_to_int(frame.lookup(op.getOperand(1), *z3));
+
+  if (ctx->check(rhs == 0 || !z3::bvsdiv_no_overflow(lhs, rhs)) == z3::sat) {
+    queue->add_failure(ctx->solver.get_model());
+  }
+  ctx->add(rhs != 0);
+  ctx->add(z3::bvsdiv_no_overflow(lhs, rhs));
+
+  frame.insert(&op, lhs % rhs);
+
+  return ExecutionResult::Continue;
+}
+
+ExecutionResult Interpreter::visitURem(llvm::BinaryOperator &op) {
+  StackFrame &frame = ctx->stack_top();
+
+  auto lhs = normalize_to_int(frame.lookup(op.getOperand(0), *z3));
+  auto rhs = normalize_to_int(frame.lookup(op.getOperand(1), *z3));
+
+  if (ctx->check(rhs == 0) == z3::sat) {
+    queue->add_failure(ctx->solver.get_model());
+  }
+  ctx->add(rhs != 0);
+
+  frame.insert(&op, lhs % rhs);
```

Comment:
Signed and unsigned remainder are different so you can't use the `%` operator for both cases. You should use the operation for unsigned remainder here. It's `z3::urem`.

It looks like the division operator has the same problem, if you could replace the unsigned division with `z3::udiv` that would be best.